packages/core/src/main/notification/NotificationService.ts

```diff
@@ -356,4 +370,82 @@ export class NotificationService extends EventEmitter {
     }
     return Decoder.fromBase64(groupId).asBytes;
   }
+
+  /**
+   * ## MLS only ##
+   * If there are pending proposals, we need to either process them,
+   * or save them in the database for later processing
+   *
+   * @param groupId groupId of the mls conversation
+   * @param delayInMs delay in ms before processing proposals
+   * @param eventTime time of the event that had the proposals
+   */
+  private async handlePendingProposals({delayInMs, groupId, eventTime}: HandlePendingProposalsParams) {
+    if (delayInMs > 0) {
+      const eventDate = new Date(eventTime);
+      const firingDate = eventDate.setTime(eventDate.getTime() + delayInMs);
+      try {
+        await this.database.storePendingProposal({
+          groupId,
+          firingDate,
+        });
+      } catch (error) {
+        this.logger.error('Could not store pending proposal', error);
+      } finally {
+        TaskScheduler.addTask({
+          task: () => this.commitPendingProposals({groupId}),
+          firingDate,
+          key: `${groupId}`,
+        });
+      }
+    } else {
+      await this.commitPendingProposals({groupId, skipDelete: true});
+    }
+  }
+
+  /**
+   * ## MLS only ##
+   * Commit all pending proposals for a given groupId
+   *
+   * @param groupId groupId of the conversation
+   * @param skipDelete if true, do not delete the pending proposals from the database
+   */
+  public async commitPendingProposals({groupId, skipDelete = false}: CommitPendingProposalsParams) {
+    const coreCryptoClient = this.coreCryptoClientProvider();
+    if (coreCryptoClient) {
```

Comment:
better check for !coreCryptoClient and throw an error

Reply:
this is a good question, should be a singleton no?

Reply:
I agree. It should be done in a separate PR tho.